File: 347. Top K Frequent Elements/solution.py
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        count = {}
        frequency_buckets = [[] for _ in range(len(nums) + 1)]

        for num in nums:
            count[num] = count.get(num, 0) + 1

        for num, count in count.items():
            frequency_buckets[count].append(num)

        topK = []

        for i in range(len(frequency_buckets) - 1, 0, -1):
            for num in frequency_buckets[i]:
                topK.append(num)
                if len(topK) == k:
                    return topK

        return topK

# Time complexity: O(N)
# Space complexity: O(N)

File: 347. Top K Frequent Elements/Solution.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> count = new HashMap<>();
        List<Integer>[] frequencyBuckets = new List[nums.length + 1];
        for (int i = 0; i < nums.length + 1; i++) {
            frequencyBuckets[i] = new ArrayList<>();
        }

        for (int num: nums) {
            count.put(num, count.getOrDefault(num, 0) + 1);
        }

        for (Map.Entry<Integer, Integer> entry: count.entrySet()) {
            int num = entry.getKey();
            int cnt = entry.getValue();
            frequencyBuckets[cnt].add(num);
        }

        int[] topK = new int[k];
        int index = 0;
        for (int i = frequencyBuckets.length - 1; i > 0; i--) {
            for (int num: frequencyBuckets[i]) {
                topK[index] = num;
                index++;
                if (index == k) {
                    return topK;
                }
            }
        }
        return topK;
    }
}

// Time complexity: O(N)
// Space complexity: O(N)

